### DISCUSSION TITLE
Test next/image src with Jest and react testing library

### DISCUSSION THREAD

**Discussion by luisorbaiceta** [vercel/next.js#32325] [Help] [answered]
### What version of Next.js are you using?

12.0.4

### What version of Node.js are you using?

16.13.1

### What browser are you using?

-

### What operating system are you using?

Windows (Docker Node.js devcontainer)

### How are you deploying your application?

-

### Describe the Bug

When testing a component that has the `next/image`

```tsx
import {FC} from 'react';
import Img, {ImageProps} from 'next/image';

type ImgProps = {
  full?: boolean;
} & ImageProps;

const FullImage: FC<ImgProps> = ({full, ...rest}) => {
  let props = {
    ...rest,
  };
  // nice defaults for full img
  if (full) {
    props = {
      objectFit: 'cover',
      layout: 'fill',
      ...props,
    };
  }
  return props.layout == 'fill' ? (
    <div className='relative h-full w-full'>
      <Img {...props} />
    </div>
  ) : (
    <Img {...props} />
  );
};

export default FullImage;
```

```tsx
it('Adds correct src attribute', () => {
  const alt = 'test-image';
  const src = '/img/mock/shoes.png'; // I have tried with a url too
  render(<Image src={src} alt={alt} width={100} height={100} />);

  const img = screen.getByAltText(alt);
  expect(img.getAttribute('src')).toBe(src);
});
```

The received `src` value is always `data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7`, for what I have seen, this value corresponds to the default applied by the [`next/image` component](https://github.com/vercel/next.js/blob/7e055dddbb5c686dcb8f07828a6bc95aa079d512/packages/next/client/image.tsx#L18) 

### Expected Behavior

I expect the source to be `/__next/image`.. So I can then add a custom loader to prevent the default transformation performed by the `next/image` component.

### To Reproduce

1. Add jest tests with next.js config as stated in the next.js docs.
2. Test a component that implements the `next/image` component.
3. Pass any src to the next/image component

**Comment by balazsorban44** [comment, marked as answer]
The Next Image component loads the image lazily, usually only after entering the viewport. You might have to wait for that or add a `priority={true}` prop on the image.

**Comment by veller** [comment]
@luisorbaiceta I'm facing the same issue but wasn't able to fix. Do you mind posting the final code? Thanks.

**Comment by gr8pathik** [comment]
I was facing the same issue and i dont think it is good to change the actual code and add priority in image tag. Instead of that we can add `waitFor`. Below is the code that worked for me without adding the priority attribute.
```
await waitFor(() => {
      expect(screen.getByTestId('imgTestId')).toHaveAttribute(
             'src',
             'expectedSrc'
      );
});
```
Dont forgot to add `async` in your test.

**Reply by YunMinSeok** [reply]
I tested this way, but I got a value like this: src="/_next/image?url=%2Fmysrc&w=828&q=75", what is the reason?

**Reply by luisorbaiceta** [reply]
I'm afraid that adding the `priority={true}` is the only way to test the url

**Reply by ryanholmesdev** [reply]
Alternativally you could mock the next/image component like this: 

```
jest.mock('next/image', () => ({
  __esModule: true,
  default: (props: any) => {
    return <img {...props} />
  },
}))
```

**Reply by ManoMylv-tBC** [reply]
Using `waitFor` worked for us, but as @YunMinSeok mentioned the src gets transformed by `Next/Link` so you can't have a straight matching test. We ended up:
1.  using the `encodeURIComponent` method on the original url.
2. using `expect.stringContaining(yourEncodedUrlHere)`  instead of `'expectedSrc'` so that it’s checking for a partial match to the value of the src.

**Reply by JayMGurav** [reply]
The below is the one that worked for me for the component with priority

component
```jsx
 <Image data-testid="image" src="/images/testImage.webp" width={150} height={40} priority alt="testImage" />
```

tests
```js

    test('to have appropriate src and srcset', () => {
        const expectedSrc = expect.stringMatching('%2Fimages%testImage.webp');
        render(
          <ImageComponent />
        );
        const image = screen.getByTestId('logo');
        expect(image).toHaveAttribute("src", expectedSrc)
        expect(image).toHaveAttribute("srcset", expectedSrc)
    });

   test('render image with appropriate alt text', () => {
        render(
          <ImageComponent />
        );
        const image = screen.getByTestId('logo');
        expect(logo).toHaveAttribute("alt", "testImage")
    })
```

however I'm not able to test the image dimensions, the attribute value of width is coming out to be `null` for some reason and the style attribute returns `0px`

**Reply by matheusggds** [reply]
You have no idea how you helped me... thank

**Comment by shubham-available** [comment]
@luisorbaiceta 

Great help, thanks.
I have one doubt, 
`I expect the source to be /__next/image.. So I can then add a custom loader to prevent the default transformation performed by the next/image component.`

how did you manage to get the exact src. I am getting something like this,
`/_next/image?url=....`

**Reply by AlvSovereign** [reply]
Use the `unoptimised` flag and set it to `true`. This will stop Next.js image optimisations and give you the exact `src` in are testing for.

**Reply by rdalfonso** [reply]
Question: My `next.config.js` file has ` images: { unoptimized: true, }` and `next/images` work in dev, they are unoptimized. However, when I run a JEST test the` next/images `are rendered as optimized. Why is this happening if I'm using the ` images: { unoptimized: true, }`?

**Comment by hamza-ashfaq751** [comment]
`jest.mock('next/image', () => ({
    __esModule: true,
    default: (
        props: JSX.IntrinsicAttributes &
            ClassAttributes<HTMLImageElement> &
            ImgHTMLAttributes<HTMLImageElement>,
    ) => <img {...props} />,
}));`

use this in every test file where Next Image is used

**Reply by the-jossie** [reply]
Worked perfectly for me

**Comment by quisido** [comment]
@hamza-ashfaq751 offers a great solution. Here is a TypeScript implementation:

```tsx
import type { ImageProps, StaticImageData } from 'next/image';
import type { ComponentType, ReactElement } from 'react';
import mapNextImageSrcToString from './map-next-image-src-to-string';

interface ESModuleDefault<T> {
  readonly __esModule: true;
  readonly default: T;
}

type StaticRequire = ImageProps['src'] extends
  | infer T
  | StaticImageData
  | string
  ? T
  : never;

const mapStaticImportToSrc = (
  staticImport: StaticImageData | StaticRequire,
): string => {
  if ('default' in staticImport) {
    return staticImport.default.src;
  }
  return staticImport.src;
};

const mapNextImageSrcToString = (
  src: StaticImageData | StaticRequire | string,
): string => {
  if (typeof src === 'string') {
    return src;
  }
  return mapStaticImportToSrc(src);
};

function MockNextImage({
  alt,
  height,
  src: nextImageSrc,
  width,
}: Readonly<ImageProps>): ReactElement {
  const imgSrc: string = mapNextImageSrcToString(nextImageSrc);
  return <img alt={alt} height={height} src={imgSrc} width={width} />;
}

jest.mock(
  'next/image',
  (): ESModuleDefault<ComponentType<ImageProps>> => ({
    __esModule: true,
    default: MockNextImage,
  }),
);
```

**Reply by theorib** [reply]
This is great, it can be encapsulated in a separate file/function and then imported into your `beforeAll()` function for example.
I've run some tests using it like that with Vitest (replacing `jest.mock` with `vi.mock`) and It works a charm.

Thank you.

**Comment by rdalfonso** [comment]
My `next.config.js` file has ` images: { unoptimized: true, }` and `next/images` work in dev, they are unoptimized. However, when I run a JEST test the` next/images` are rendered as _optimized_. Why is this happening if I'm using the  ` images: { unoptimized: true, }`?

**Comment by vvu060** [comment]
```
  // Encode the original URL
    const encodedUrl = encodeURIComponent('https://example.com/avatar.jpg');

    // Verify that the image has the correct source and alt text
    expect(imgElement).toHaveAttribute(
      'src',
      `/_next/image?url=${encodedUrl}&w=96&q=75`
    );
    
    ```
